torchfl/cli.py

```diff
@@ -13,28 +44,142 @@ def cli_parser() -> Namespace:
+        help="name of the experiment.",
+    )
+    general.add_argument(
+        "--experiment_version",
+        type=float,
+        default=0.1,
+        help="version of the experiment. Default: 0.1",
+    )
+    general.add_argument("-v", "--verbose", action="store_true")
+
+    # datamodule args
+    datamodule.add_argument(
+        "--data_dir",
+        type=str,
+        help=f"path to store the downloaded data-related files. Default: {TORCHFL_DIR}",
+        default=TORCHFL_DIR,
+    )
+    datamodule.add_argument(
+        "--dataset_group",
+        type=str,
+        choices=DATASET_GROUPS_MAP.keys(),
+        help="name of the dataset group.",
+    )
+    datamodule.add_argument(
+        "--dataset_name",
+        type=str,
+        action=DatasetNameSanitizer,
+        help="name of the dataset (should belong to a data group).",
+    )
+    datamodule.add_argument(
+        "--validation_split",
+        type=float,
+        help="validation split ratio. Default: 0.1",
+        default=0.1,
+    )
+    datamodule.add_argument(
+        "--train_bs",
+        type=int,
+        help="batch size used for the training dataset. Default: 128",
+        default=128,
+    )
+    datamodule.add_argument(
+        "--validation_bs",
+        type=int,
+        help="batch size used for the validation dataset. Default: 1",
+        default=1,
+    )
+    datamodule.add_argument(
+        "--test_bs",
+        type=int,
+        help="batch size used for the testing dataset. Default: 1",
+        default=1,
+    )
+    datamodule.add_argument(
+        "--predict_bs",
+        type=int,
+        help="batch size used for the prediction dataset. Default: 1",
+        default=1,
+    )
+    # NOTE: we can't support the following args as of now: train_transforms, val_transforms, test_transforms, predict_transforms
+
+    # model args
+    model.add_argument(
+        "--model_name",
+        type=str,
+        required=("--config" not in sys.argv),
+        help="name of the model.",
+    )
+    # FIXME: Write sanitizer validation functions for all of the following arguments
+    model.add_argument("--num_classes", type=int, help="number of classes.")
+    model.add_argument("--num_channels", type=int, help="number of channels.")
+    model.add_argument(
+        "--activation_fn", type=str, help="activation function name."
+    )
+    model.add_argument("--pretrained", action="store_false")
+    model.add_argument("--feature_extract", action="store_false")
+
+    # optimizer args
+    # FIXME: Write sanitizer validation functions for all of the following arguments
+    optimizer.add_argument(
+        "--optimizer_name",
+        type=str,
+        required=("--config" not in sys.argv),
+        help="name of the optimizer.",
+    )
+    optimizer.add_argument("--lr", type=float, help="learning rate.")
+    # NOTE: Optimizers support many more arguments. We can't support all of them in CLI, but we will add support for them in YAML configs.
```

Comment:
We could write this so that the CLI and YAML are treated as the same thing and are abstracted to some sort of intermediate representation. Right now you are using the Namespaces class in argparse which we could use as the argparse backend and then write a translator.. This could make so that in the future you can load the config through anything that you want even a unix pipe or something. Ill add this to the issue when I get a chance.

Reply:
That's definitely a good idea. Let's put more thought into this and details into the issue created here - https://github.com/torchfl-org/torchfl/issues/23. As I said, I haven't thought this thing through.